can/zynqmp-can/zynqmp-can.c

```diff
@@ -0,0 +1,552 @@
+		can.base->ICR = (1 << 7);
+
+		/* Check if there is still data in the RX FIFO */
+		uint32_t isr = can.base->ISR;
+		if (0 == (isr & (1 << 7))) {
+			break;
+		}
+
+		/* Read a frame from the FIFO */
+		buf[i].id = (can.base->RXFIFO_ID >> 21) & 0x7FF;
+		buf[i].len = (can.base->RXFIFO_DLC >> 28) & 0xF;
+		buf[i].payload.words[0] = swap_endianness(can.base->RXFIFO_DATA1);
+		buf[i].payload.words[1] = swap_endianness(can.base->RXFIFO_DATA2);
+		(*recvFrames)++;
+	}
+
+	return EOK;
+}
+
+/**
+ * Main thread for handling CAN operations.
+ * Processes incoming messages and performs the requested operations.
+ */
+static void can_thread(void *arg)
+{
+	msg_t msg;
+	msg_rid_t rid;
+	uint32_t port = can.oid.port;
+
+	mutexLock(can.lock);
+
+	for (;;) {
+		/* Wait indefinitely for a message */
+		if (msgRecv(port, &msg, &rid) < 0) {
+			continue;
+		}
+		switch (msg.type) {
+			/* This driver supports only one way of communication with it */
+			case mtDevCtl: {
+				/* Decode extra information passed with the request */
+				zynqmp_canDriverReq *req = (zynqmp_canDriverReq *)msg.i.raw;
+				zynqmp_canDriverResp *resp = (zynqmp_canDriverResp *)msg.o.raw;
+
+				/* Choose operation based on opcode */
+				if (req->operation == zynqmp_can_opSend) {
+					zynqmp_canFrame *frames = (zynqmp_canFrame *)msg.i.data;
+					uint32_t count = msg.i.size / sizeof(zynqmp_canFrame);
+					msg.o.err = send_frames(frames, count, req->blockTxOperation, &resp->framesReceivedSend);
+				}
+				else if (req->operation == zynqmp_can_opRecv) {
+					zynqmp_canFrame *buf = (zynqmp_canFrame *)msg.o.data;
+					uint32_t bufLen = msg.o.size / sizeof(zynqmp_canFrame);
+					msg.o.err = recv_frames(buf, bufLen, req->timeoutUs, &resp->framesReceivedSend);
+				}
+				else {
+					printf("zynqmp-can: unknown opcode\n");
+					msg.o.err = -ENOSYS;
+				}
+			} break;
+
+			case mtOpen:
+			case mtClose:
+			case mtWrite:
+			case mtRead:
+			case mtGetAttr:
+			default: {
+				msg.o.err = -ENOSYS;
+				break;
+			}
+		}
+		msgRespond(port, &msg, rid);
+	}
+
+	mutexUnlock(can.lock);
+}
+
+/**
+ * Main entry point for the CAN driver.
+ * Parses command-line arguments, initializes the CAN peripheral, and starts the message handling thread.
+ */
+int main(int argc, char **argv)
+{
+	int c = 0;
+
+	/* Default arguments */
+	int can_id = can_1;
+	int baudrate = CAN_BAUDRATE_KBPS_DEFAULT;
+
+	/* Parse command-line arguments */
+	if (argc > 1) {
+		while ((c = getopt(argc, argv, "n:b:h")) != -1) {
+			switch (c) {
+				case 'b': {
+					baudrate = atoi(optarg);
+					/* Find the matching time configuration */
+					bool timeCfgFound = false;
+					for (uint32_t i = 0; i < ARRAY_LENGTH(timeCfg); i++) {
+						if (timeCfg[i].baudrate_kbps == baudrate) {
+							timeCfgFound = true;
+						}
+					}
+					if (!timeCfgFound) {
+						debug("zynqmp-can: failed to configure desired baud rate\n");
+						return EXIT_FAILURE;
+					}
+				} break;
```

Comment:
Redundant, already checked in can_init

```suggestion
				case 'b':
					baudrate = atoi(optarg);
					break;
```